src/tfr_executive/src/clock_service.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * The clock service is a really simple node that exposes some utility services
+ * to simplify the operation of the executive system.
+ * 
+ * PARAMETERS 
+ * - ~driving_time: The amount of time we anticipate driving, to take in 
+ *                  seconds. (`float`, default 60)
+ * - ~dumping_time: The amount of time we anticipate dumping to take in 
+ *                  seconds. (`float`, default 30)
+ * - ~mission_time: The amount of time we allocate for the mission in total 
+ *                  seconds. (`float`, default: 600)
+ * SERVICES  
+ * 1. start_mission: starts the mission clock.
+ * 2. time_remaining: returns the amount of time remaining in the mission
+ * 3. digging_time: Gives the dumping time which is equal to  
+ *                  (duration - start - driving_time - dumping_time) or 0 
+ *                  if negative duration. 
+ * */
+#include <ros/ros.h>
+#include <tfr_msgs/EmptySrv.h>
+#include <tfr_msgs/DurationSrv.h>
+
+class ClockService
+{
+    public:
+        ClockService(ros::NodeHandle &n, ros::Duration& mission, 
+                ros::Duration& driving, ros::Duration& dumping):
+            start_mission{n.advertiseService("start_mission", &ClockService::startMission, this)},
+            time_remaining{n.advertiseService("time_remaining", &ClockService::timeRemaining, this)},
+            digging_time{n.advertiseService("digging_time", &ClockService::diggingTime , this)},
+            mission_start{},
+            mission_duration{mission},
+            driving_duration{driving},
+            dumping_duration{dumping}
+        {
+        }
+
+        ~ClockService() = default;
+        ClockService(const ClockService&) = delete;
+        ClockService& operator=(const ClockService&) = delete;
+        ClockService(ClockService&&) = delete;
+        ClockService& operator=(ClockService&&) = delete;
+
+
+    private: 
+        /*
+         * Starts the mission clock, takes in the empty service request.
+         * */
+        bool startMission(tfr_msgs::EmptySrv::Request &req,
+                tfr_msgs::EmptySrv::Response &res)
+        {
+            ROS_INFO("mission started");
+            mission_start = ros::Time::now();
+            return true;
+        }
+
+        /*
+         * Gives the remaining time in the mission, fills a duration request.
+         * */
+        bool timeRemaining(tfr_msgs::DurationSrv::Request &req,
+                tfr_msgs::DurationSrv::Response &res)
+        {
+            res.duration = mission_duration - (ros::Time::now() - mission_start);
+            return true;
+        }
+
+        /*
+         * Gives the time to allow for digging, fills a duration request.
+         * */
+        bool diggingTime(tfr_msgs::DurationSrv::Request &req,
+                tfr_msgs::DurationSrv::Response &res)
+        {
+            timeRemaining(req, res);
+            if (mission_start.toSec() == 0.0)
```

Comment:
Not sure if this is an issue here, but == with a double?

Reply:
You are correct. Well we are determining if `start_clock` hasn't been set yet. The first thing I do is call `mission_start`'s default parameterless constructor, which __should__ zero initialize POD fields like time. 
This article's second response show what the ISO standards committee says about that.
https://stackoverflow.com/questions/3803153/what-are-primitive-types-default-initialized-to-in-c
This article describes what the standard defines for object initialization. So this is safe for this use case provided, as long as they continue to follow the rules.

However it would be better to initialize this to something that doesn't rely on open source programmers following the rules. So I'll initialize it to `nan` and check for `nan` instead.

Reply:
Perfect, alrighty.

Reply:
looks like ros::Time provides a convenience function just for this purpose. I'll use that instead.